src/server/auth.js

```diff
@@ -4,19 +4,117 @@
+        req.iauth.grant = req.session[IMPERSONATION_TOKEN_KEY];
+    }
+    next(req, res);
+};
+
+/**
+ * Obtains the user's actual access token from the request.
+ * @param {Object} req
+ */
+export const getActualToken = (req) => {
+    return req?.kauth?.grant?.access_token;
+};
+
+/**
+ * Obtains the access token that is currently in effect if any are in effect.
+ * Some administrators have the ability to impersonate other users for
+ * troubleshooting purposes. If impersonation is in effect then the effective
+ * token is the one that was obtained in order to impersonate another user.
+ * Otherwise, the effective token is the user's actual token. If the user is
+ * not logged in at all then a null value will be returned.
+ *
+ * @param {Object} req
+ */
+export const getEffectiveToken = (req) => {
+    return req?.iauth?.grant?.access_token || getActualToken(req);
+};
+
 /**
  * Extracts the username from the calims in the JWT access token.
  *
  * @param {Object} req
  */
 export const getUserID = (req) => {
-    return req?.kauth?.grant?.access_token?.content?.preferred_username;
+    return getEffectiveToken(req)?.content?.preferred_username;
+};
+
+/**
+ * Attempts to obtain an impersonation token for the authenticated user. The
+ * user (obviously) must be logged in for this to work, but This function makes
+ * no attempt to verify that this is the case. Instead, the calling handler
+ * must ensure that the user is logged in. The impersonation token will only be
+ * granted if the user has impersonation privileges in Keycloak.
+ *
+ * @param {Object} req
+ * @param {Object} res
+ * @param {Function} next
+ * @param {String} username
+ */
+export const impersonate = (req, res, next, username) => {
+    // Build the URL for the token exchange.
+    const url = new URL(config.keycloakServerURL);
+    url.pathname = path.join(
+        url.pathname,
+        "realms",
+        config.keycloakRealm,
+        "protocol",
+        "openid-connect",
+        "token"
+    );
+
+    // Build the request body.
+    const data = {
+        grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
+        client_id: config.keycloakClientID,
+        client_secret: config.keycloakClientSecret,
+        subject_token: getActualToken(),
+        requested_token_type: "urn:ietf:params:oauth:token-type:access_token",
+        requested_subject: username,
+    };
+
+    // Send the request.
+    return axiosInstance
+        .post(url.toString(), qs.stringify(data))
+        .then((response) => {
+            console.log(response.data);
+            req.session[IMPERSONATION_TOKEN_KEY] = response.data;
+        })
+        .then(next(req, res))
+        .catch(() => next(req, res));
+};
+
+/**
+ * Stops impersonation.
+ * @param {Object} req
+ * @param {Object} res
+ * @param {Function} next
+ */
+export const stopImpersonation = (req, res, next) => {
+    delete req.session[IMPERSONATION_TOKEN_KEY];
```

Comment:
Not really a comment on your PR, but I had to look this up.  I had no idea a simple `delete` existed :smile: The syntax looks so strange to me, I thought it was pseudo code